### PR TITLE
Support Iceberg nested and binary GPU writes

### LINKED ISSUE #14239: [FEA] Support write nested data type for iceberg.
Part of #12298 , in this issue we support writing to iceberg.

### PR DESCRIPTION
**Pull request by res-life**
Fixes #14239.

### Depends on:
* https://github.com/rapidsai/cudf/pull/22143
* https://github.com/rapidsai/cudf/pull/22188
* https://github.com/rapidsai/cudf/pull/22422

### Description

- Preserve Iceberg field ids and the original write schema so nested arrays, maps, and binary columns keep the Parquet layout and metrics Iceberg expects.
- Extend Iceberg integration coverage for append, CTAS, RTAS, and overwrite flows, including focused regression coverage for binary and array<binary> writes.
- Update ITs

### Checklists

Documentation
- [ ] Updated for new or modified user-facing features or behaviors
- [x] No user-facing change

Testing
- [x] Added or modified tests to cover new code paths
- [ ] Covered by existing tests
      (Please provide the names of the existing tests in the PR description.)
- [ ] Not required

Performance
- [ ] Tests ran and results are added in the PR description
- [ ] Issue filed with a link in the PR description
- [x] Not required

### CODE REVIEW

File: sql-plugin/src/main/spark350/scala/org/apache/spark/sql/execution/datasources/v2/GpuMergeRowsExec.scala
Comment (liurenjie1024) [inline sql-plugin/src/main/spark350/scala/org/apache/spark/sql/execution/datasources/v2/GpuMergeRowsExec.scala:1]: Is this change related?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/IcebergParquetMetadataKeys.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package com.nvidia.spark.rapids
+
+/**
+ * Shared Spark metadata keys used to round-trip nested Iceberg Parquet field ids through Spark
+ * schemas and into cuDF writer options.
+ */
+object IcebergParquetMetadataKeys {
+  val LIST_ELEMENT_FIELD_ID_METADATA_KEY =
```
Comment (liurenjie1024) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/IcebergParquetMetadataKeys.scala]: IIRC, we have `FIELD_ID_METADATA_KEY`, why not use it?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala [hunk 2]
```diff
@@ -36,6 +36,57 @@ import org.apache.spark.sql.types._
...
   // Parquet field ID metadata key
   private val FIELD_ID_METADATA_KEY = "parquet.field.id"
+  import IcebergParquetMetadataKeys._
+
+  // Reflection-based access to private fields on cuDF's ColumnWriterOptions.
+  // cuDF's Java list/map builder path drops the container's parquet field id and has no
+  // public setter for `isBinary` on a built options object, so we reach into the private
+  // fields here. TODO(cuDF): remove once public setters are exposed upstream.
+  // Tracking issue: https://github.com/rapidsai/cudf/issues/NEW (file before merging).
+  private def accessibleField(name: String): java.lang.reflect.Field = {
```
Comment (liurenjie1024) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala]: We should not use this, please file a pr in cudf first.

File: integration_tests/src/main/python/iceberg/iceberg_append_test.py [hunk 3]
```diff
@@ -194,30 +252,98 @@ def test_insert_into_partitioned_table_full_coverage(spark_tmp_table_factory, pa
...
+
+    cpu_data = with_cpu_session(lambda spark: spark.table(cpu_table_name).collect())
+    gpu_data = with_cpu_session(lambda spark: spark.table(gpu_table_name).collect())
+    assert_equal_with_local_sort(cpu_data, gpu_data)
+
+
+@iceberg
+@ignore_order(local=True)
+@pytest.mark.skipif(is_iceberg_remote_catalog(), reason="Skip for remote catalog to reduce test time")
+def test_insert_into_partitioned_table_all_cols(spark_tmp_table_factory, spark_tmp_path):
```
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/iceberg_append_test.py]: Is test_insert_into_partitioned_table_all_cols and test_insert_into_partitioned_table_nested_types duplicated?

File: integration_tests/src/main/python/iceberg/iceberg_ctas_test.py [hunk 4]
```diff
@@ -215,51 +232,94 @@ def run_ctas(spark):
...
+                                assert_no_cpu_project=True)
 
 
 @iceberg
 @ignore_order(local=True)
-@allow_non_gpu('AtomicCreateTableAsSelectExec', 'AppendDataExec', 'ShuffleExchangeExec', 'ProjectExec')
 @pytest.mark.skipif(is_iceberg_remote_catalog(), reason="Skip for remote catalog to reduce test time")
 @allow_non_gpu_conditional(is_spark_400_or_later(), "EmptyRelationExec")
-def test_ctas_partitioned_table_all_cols_fallback(spark_tmp_table_factory):
+def test_ctas_partitioned_table_all_cols(spark_tmp_table_factory, spark_tmp_path):
```
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/iceberg_ctas_test.py]: Ditto, I think they have some duplicates with prev tests?

File: integration_tests/src/main/python/iceberg/iceberg_delete_test.py
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/iceberg_delete_test.py]: This should be removed?

Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/iceberg_delete_test.py]: This should be removed?

File: iceberg/common/src/main/scala/org/apache/iceberg/parquet/GpuParquetMetricsHelper.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.stream.{Stream => JStream}
+
+import org.apache.iceberg.{FieldMetrics, Metrics, MetricsConfig, Schema}
+import org.apache.iceberg.shaded.org.apache.parquet.hadoop.metadata.ParquetMetadata
+
+/**
+ * Helper in Iceberg's parquet package so GPU code can reuse ParquetMetrics with the original
+ * Iceberg schema instead of reconstructing schema from the written Parquet footer.
+ */
+object GpuParquetMetricsHelper {
```
Comment (liurenjie1024) [inline iceberg/common/src/main/scala/org/apache/iceberg/parquet/GpuParquetMetricsHelper.scala]: Do we really need? I think if we correct set field id in parquet footer, we don't need to pass the original iceberg schema.

File: iceberg/common/src/main/scala/org/apache/iceberg/spark/GpuTypeToSparkType.scala [hunk 6]
```diff
@@ -45,6 +46,40 @@ object GpuTypeToSparkType {
 
     builder.build()
   }
+
+  private def nestedMetadataJson(fieldType: org.apache.iceberg.types.Type): Option[String] = {
```
Comment (liurenjie1024) [inline iceberg/common/src/main/scala/org/apache/iceberg/spark/GpuTypeToSparkType.scala]: I would like to see some ut and doc about this part, it a little hacky.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/IcebergParquetMetadataKeys.scala (hunk 1 above)
Comment (liurenjie1024) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/IcebergParquetMetadataKeys.scala]: I think I got your point, we should put this to GpuTypeToSparkType?

**Review by liurenjie1024 — COMMENTED** [review]
Thanks @res-life for this pr. One missing point is that I hope you can do a careful check in the dml tests, since there are some fallbacks no longer necessary.

**Comment by res-life** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala (hunk 2 above)
Comment (res-life) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala]: Done: https://github.com/rapidsai/cudf/issues/22347.

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

File: docs/additional-functionality/advanced_configs.md
Comment (liurenjie1024) [inline docs/additional-functionality/advanced_configs.md:1]: Seems unrelated?

File: docs/dev/nvtx_ranges.md
Comment (liurenjie1024) [inline docs/dev/nvtx_ranges.md:1]: Seems unrelated?

File: docs/configs.md
Comment (liurenjie1024) [inline docs/configs.md:1]: Ditto.

File: docs/supported_ops.md
Comment (liurenjie1024) [inline docs/supported_ops.md:1]: Ditto.

File: tests/src/test/spark350/scala/org/apache/iceberg/spark/GpuTypeToSparkTypeSuite.scala [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+  test("nestedMetadataJson for a flat list records only the element id") {
+    val listType = Types.ListType.ofOptional(11, Types.IntegerType.get())
+    val md = parse(GpuTypeToSparkType.nestedMetadataJson(listType))
+
+    assert(md.getLong(LIST_ELEMENT_FIELD_ID_METADATA_KEY) == 11L)
+    assert(!md.contains(LIST_ELEMENT_NESTED_IDS_METADATA_KEY))
+  }
+
+  test("nestedMetadataJson for a flat map records only the key/value ids") {
```
Comment (liurenjie1024) [inline tests/src/test/spark350/scala/org/apache/iceberg/spark/GpuTypeToSparkTypeSuite.scala]: For these ut, could you add check for the json string? It would be easier to understand.

File: integration_tests/src/main/python/iceberg/__init__.py [hunk 8]
```diff
@@ -71,6 +72,105 @@
...
+]
+
+# Nested types intended for focused positive GPU Iceberg tests.
+#
+# Covers nullable nested fields, empty arrays/maps, and NaN/Inf floats/doubles inside nested
+# containers so merge/update/delete/append/CTAS/RTAS/overwrite nested_types tests exercise the
+# same edge cases that `iceberg_full_gens_list`-based `_all_cols` tests do. `_c0` and `_c1` stay
+# strict non-nullable IntegerGen/LongGen so merge/update/delete join keys and predicates don't
+# get trivialized by nulls.
+iceberg_nested_write_gens_list = ([IntegerGen(nullable=False), LongGen(nullable=False),
```
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/__init__.py]: This seems has a lot duplication with `iceberg_full_gens_list`? Our tests has been reported to be quite slow, so would be better  if we could reduce unnecessary duplication and only add ones that are needed.

File: integration_tests/src/main/python/iceberg/__init__.py [hunk 9]
```diff
@@ -71,6 +72,105 @@
                                               ['child2', int_gen]]))] +
                           iceberg_map_gens + decimal_gens)
 
+# Scalar top-level column generators.
+#
+# Top-level columns are non-nullable so rows are always present (row-level null coverage lives in
```
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/__init__.py]: Why top level columns have to be non-nullable?

File: integration_tests/src/main/python/iceberg/iceberg_append_test.py [hunk 10]
```diff
@@ -102,27 +105,42 @@ def insert_data(spark, table_name: str):
...
         view_name = spark_tmp_table_factory.get()
         df.createOrReplaceTempView(view_name)
         return spark.sql(f"INSERT INTO {table_name} SELECT * FROM {view_name}")
 
-    table_name = get_full_table_name(spark_tmp_table_factory)
-    create_iceberg_table(table_name, table_prop=table_prop, df_gen=this_gen_df)
+    def insert_data_and_assert_gpu_plan(spark, table_name: str):
+        df = insert_data(spark, table_name)
+        df.collect()
+        assert_no_cpu_project_exec(spark, df)
```
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/iceberg_append_test.py]: This is for debugging only?

File: integration_tests/src/main/python/iceberg/iceberg_append_test.py [hunk 11]
```diff
@@ -102,27 +105,42 @@ def insert_data(spark, table_name: str):
...
 @iceberg
 @ignore_order(local=True)
-@allow_non_gpu('AppendDataExec', 'ShuffleExchangeExec', 'ProjectExec')
 @pytest.mark.skipif(is_iceberg_remote_catalog(), reason="Skip for remote catalog to reduce test time")
-def test_insert_into_unpartitioned_table_all_cols_fallback(spark_tmp_table_factory):
+def test_insert_into_unpartitioned_table_all_cols(spark_tmp_table_factory, spark_tmp_path):
     table_prop = {"format-version": "2"}
+    cols = [f"_c{idx}" for idx, _ in enumerate(iceberg_full_gens_list)]
+    gen_list = list(zip(cols, iceberg_full_gens_list))
+    source_path = materialize_parquet_source(spark_tmp_path, gen_list)
```
Comment (liurenjie1024) [inline integration_tests/src/main/python/iceberg/iceberg_append_test.py]: This is debugging only codes?

**Comment by res-life** [comment]
build

File: docs/additional-functionality/advanced_configs.md
Comment (res-life) [reply docs/additional-functionality/advanced_configs.md:1]: The doc is generated via `./build/buildall --profile=330,331,332,333,334,340,341,342,343,344,350,351,352,353,354,355,356,357,358 -P=4 && ./build/buildall --scala213 --profile=400,401,402,411 -P=4`
It's werid, the premerge passed. Merged the main branch, build again to check.

File: docs/dev/nvtx_ranges.md
Comment (res-life) [reply docs/dev/nvtx_ranges.md:1]: The doc is generated via `./build/buildall --profile=330,331,332,333,334,340,341,342,343,344,350,351,352,353,354,355,356,357,358 -P=4 && ./build/buildall --scala213 --profile=400,401,402,411 -P=4`
It's werid, the premerge passed. Merged the main branch, build again to check.

File: docs/configs.md
Comment (res-life) [reply docs/configs.md:1]: The doc is generated via `./build/buildall --profile=330,331,332,333,334,340,341,342,343,344,350,351,352,353,354,355,356,357,358 -P=4 && ./build/buildall --scala213 --profile=400,401,402,411 -P=4`
It's werid, the premerge passed. Merged the main branch, build again to check.

File: docs/supported_ops.md
Comment (res-life) [reply docs/supported_ops.md:1]: The doc is generated via `./build/buildall --profile=330,331,332,333,334,340,341,342,343,344,350,351,352,353,354,355,356,357,358 -P=4 && ./build/buildall --scala213 --profile=400,401,402,411 -P=4`
It's werid, the premerge passed. Merged the main branch, build again to check.

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
All the Iceberg cases passed locally:
```
= 853 passed, 15 skipped, 3 xfailed, 2 xpassed, 101 warnings in 4491.24s (1:14:51) =
```

**Comment by res-life** [comment]
Premerge failed two cases, not related to this PR.
Error is:
```
Cause: ai.rapids.cudf.CudfException: CUDF failure at: ../../../thirdparty/cudf/cpp/src/io/comp/nvcomp_adapter.cpp:565: nvCOMP error: nvcompErrorInternal
```

**Comment by res-life** [comment]
build

File: iceberg/common/src/main/scala/org/apache/iceberg/spark/GpuTypeToSparkType.scala [hunk 12]
```diff
@@ -45,6 +46,63 @@ object GpuTypeToSparkType {
...
+        val fields = map.fields().asScala
+        val keyField = fields.head
+        val valueField = fields(1)
+        builder.putLong(MAP_KEY_FIELD_ID_METADATA_KEY, keyField.fieldId())
+        nestedMetadataJson(keyField.`type`())
+          .foreach(builder.putString(MAP_KEY_NESTED_IDS_METADATA_KEY, _))
+        builder.putLong(MAP_VALUE_FIELD_ID_METADATA_KEY, valueField.fieldId())
+        nestedMetadataJson(valueField.`type`())
+          .foreach(builder.putString(MAP_VALUE_NESTED_IDS_METADATA_KEY, _))
+      case _ =>
```
Comment (liurenjie1024) [inline iceberg/common/src/main/scala/org/apache/iceberg/spark/GpuTypeToSparkType.scala]: What if the nested type is a struct? Say `List<List<Struct>`?

File: iceberg/common/src/main/scala/org/apache/iceberg/spark/GpuTypeToSparkType.scala [hunk 13]
```diff
@@ -55,7 +113,9 @@ class GpuTypeToSparkType extends TypeToSparkType {
       .zip(fieldResults.asScala)
       .map {
         case (field, fieldResult) =>
-          val metadata = fieldMetadataOf(field.fieldId())
+          val metadataBuilder = new MetadataBuilder().withMetadata(fieldMetadataOf(field.fieldId()))
+          GpuTypeToSparkType.appendNestedFieldIdMetadata(metadataBuilder, field.`type`())
```
Comment (liurenjie1024) [inline iceberg/common/src/main/scala/org/apache/iceberg/spark/GpuTypeToSparkType.scala]: This travel the data type in a top down recursive approach. In fact, the `TypeToSparkType` is a post order traversal of iceberg`s `Schema` data structure, and in my opinion you could utilize this feature to build a json representation of nested field ids only for top level fields.

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala (hunk 2 above)
Comment (liurenjie1024) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/SchemaUtils.scala]: rapidsai/cudf#22347 is closed, should we remove the reflection now?

File: integration_tests/src/main/python/iceberg/iceberg_append_test.py (hunk 10 above)
Comment (liurenjie1024) [reply integration_tests/src/main/python/iceberg/iceberg_append_test.py]: Hi, this is not resolved.

Comment (liurenjie1024) [reply integration_tests/src/main/python/iceberg/iceberg_append_test.py]: I don't understand why we need this.

File: integration_tests/src/main/python/iceberg/iceberg_append_test.py (hunk 11 above)
Comment (liurenjie1024) [reply integration_tests/src/main/python/iceberg/iceberg_append_test.py]: Hi, this is not resolved.

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM, but please resolve the comments.

**Comment by res-life** [comment]
- fixed reflection via: [5036c4f](https://github.com/NVIDIA/spark-rapids/pull/14611/commits/5036c4fb513c4bad288a018126eed8d95d472048)
- about `assert_no_cpu_project_exec`, it asserts the GPU plan contains no CPU ProjectExec. assert_no_cpu_project=True → the helper additionally asserts no CPU ProjectExec in the GPU plan. 
E.g.: 
```python
_assert_gpu_equals_cpu_ctas(spark_tmp_table_factory,                          
......                                                                                        
                                  assert_no_cpu_project=True)
```
If assert_no_cpu_project=False, the _assert_gpu_equals_cpu_ctas will success if the CPU result and GPU result are same. `assert_no_cpu_project=True` add an additional check: verify there is no project falls back to CPU in GPU plan.

**Comment by res-life** [comment]
- [x] fixed reflection: done
- [x] materialize_parquet_source: done
- [x] assert_no_cpu_project_exec: done
Passed locally for all the Iceberg ITs on Spark 356:
= 853 passed, 15 skipped, 2 xfailed, 3 xpassed, 96 warnings in 4126.19s (1:08:46) =

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
CI failed by known issue: 
14609: [BUG] hash_aggregate_test.py::test_hash_grpby_sum failed by shuffle timeout 120s (2026-04-15T05:34:37Z)

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
build

**Review by liurenjie1024 — APPROVED** [review]

File: docs/additional-functionality/advanced_configs.md
Comment (thirtiseven) [reply docs/additional-functionality/advanced_configs.md:1]: Recently, the doc check has been broken: https://github.com/NVIDIA/spark-rapids/issues/14767. For `./build/buildall --scala213`, it generates a doc based on 400. I usually run `./build/buildall --scala213 && ./build/buildall` to generate all the files, so the doc is based on 330.

We should bring back the 330 docs.
